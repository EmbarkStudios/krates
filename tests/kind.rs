mod util;

use util::build;

#[test]
fn all_the_things() {
    let grafs = build("all-features.json", krates::Builder::new()).unwrap();

    assert!(grafs.filtered.is_empty());
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn filters_dev() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        // This shouldn't actually affect anything, as dev dependencies
        // for non-workspace crates are already not resolved
        assert!(grafs.filtered.is_empty());
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);

        // This will be equivalent to to filtering workspace dev crates
        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_build() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_normal() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_build_and_dev() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::Workspace);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn only_b() {
    let mut kb = krates::Builder::new();
    kb.include_workspace_crates(["/home/jake/code/krates/tests/ws/b/Cargo.toml"]);

    let grafs = build("all-features.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn filters_after_build() {
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        let filtered = grafs.actual.krates_filtered(krates::DepKind::Dev);

        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();

        let expected = format!("{:#?}", grafs.actual.krates().collect::<Vec<_>>());
        let actual = format!("{filtered:#?}");

        similar_asserts::assert_eq!(expected, actual);
    }

    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        let filtered = grafs.actual.krates_filtered(krates::DepKind::Build);

        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();

        let expected = format!("{:#?}", grafs.actual.krates().collect::<Vec<_>>());
        let actual = format!("{filtered:#?}");

        similar_asserts::assert_eq!(expected, actual);
    }
}
