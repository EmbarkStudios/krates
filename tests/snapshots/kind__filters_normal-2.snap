---
source: tests/kind.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate a 0.1.0 (path+file:///krates/tests/ws/a)" ]
    1 [ label = "crate b 0.1.0 (path+file:///krates/tests/ws/b)" ]
    2 [ label = "crate bumpalo 3.11.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    3 [ label = "crate c 0.1.0 (path+file:///krates/tests/ws/c)" ]
    4 [ label = "crate cc 1.0.73 (git+https://github.com/alexcrichton/cc-rs#53fb72c87e5769a299f1886ead831901b9c775d6)" ]
    5 [ label = "crate cc 1.0.73 (registry+https://github.com/rust-lang/crates.io-index)" ]
    6 [ label = "crate cfg-if 1.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    7 [ label = "crate difference 2.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    8 [ label = "crate js-sys 0.3.60 (registry+https://github.com/rust-lang/crates.io-index)" ]
    9 [ label = "crate libc 0.2.133 (registry+https://github.com/rust-lang/crates.io-index)" ]
    10 [ label = "crate log 0.4.17 (registry+https://github.com/rust-lang/crates.io-index)" ]
    11 [ label = "crate once_cell 1.15.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    12 [ label = "crate proc-macro2 1.0.44 (registry+https://github.com/rust-lang/crates.io-index)" ]
    13 [ label = "crate quote 1.0.21 (registry+https://github.com/rust-lang/crates.io-index)" ]
    14 [ label = "crate ring 0.16.20 (registry+https://github.com/rust-lang/crates.io-index)" ]
    15 [ label = "crate spin 0.5.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    16 [ label = "crate syn 1.0.101 (registry+https://github.com/rust-lang/crates.io-index)" ]
    17 [ label = "crate unicode-ident 1.0.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    18 [ label = "crate untrusted 0.7.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    19 [ label = "crate wasm-bindgen 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    20 [ label = "crate wasm-bindgen-backend 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    21 [ label = "crate wasm-bindgen-macro 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    22 [ label = "crate wasm-bindgen-macro-support 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    23 [ label = "crate wasm-bindgen-shared 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    24 [ label = "crate web-sys 0.3.60 (registry+https://github.com/rust-lang/crates.io-index)" ]
    25 [ label = "crate winapi 0.3.9 (registry+https://github.com/rust-lang/crates.io-index)" ]
    26 [ label = "crate winapi-i686-pc-windows-gnu 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    27 [ label = "crate winapi-x86_64-pc-windows-gnu 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    28 [ label = "feature default" ]
    29 [ label = "feature default" ]
    30 [ label = "feature default" ]
    31 [ label = "feature default" ]
    32 [ label = "feature proc-macro" ]
    33 [ label = "feature std" ]
    34 [ label = "feature Crypto" ]
    35 [ label = "feature Window" ]
    36 [ label = "feature ntsecapi" ]
    37 [ label = "feature wtypesbase" ]
    38 [ label = "feature proc-macro" ]
    39 [ label = "feature spans" ]
    40 [ label = "feature default" ]
    41 [ label = "feature default" ]
    42 [ label = "feature default" ]
    43 [ label = "feature default" ]
    44 [ label = "feature full" ]
    45 [ label = "feature default" ]
    46 [ label = "feature spans" ]
    47 [ label = "feature visit" ]
    48 [ label = "feature spans" ]
    49 [ label = "feature leftpad" ]
    50 [ label = "feature leftier-strings" ]
    51 [ label = "feature lazy_static" ]
    52 [ label = "feature alloc" ]
    53 [ label = "feature race" ]
    54 [ label = "feature once_cell" ]
    55 [ label = "feature dev_urandom_fallback" ]
    56 [ label = "feature alloc" ]
    57 [ label = "feature quote" ]
    58 [ label = "feature proc-macro" ]
    59 [ label = "feature printing" ]
    60 [ label = "feature parsing" ]
    61 [ label = "feature derive" ]
    62 [ label = "feature clone-impls" ]
    63 [ label = "feature std" ]
    64 [ label = "feature spans" ]
    65 [ label = "feature EventTarget" ]
    0 -> 28 [ label = "(dev)" ]
    0 -> 28 [ label = "(build) 'cfg(target_os = \"linux\")'" ]
    1 -> 29 [ label = "(dev) 'cfg(target_arch = \"x86_64\")'" ]
    3 -> 4 [ label = "(build)" ]
    3 -> 30 [ label = "(dev)" ]
    8 -> 31 [ label = "" ]
    10 -> 6 [ label = "" ]
    12 -> 17 [ label = "" ]
    13 -> 12 [ label = "" ]
    13 -> 32 [ label = "" ]
    14 -> 5 [ label = "(build)" ]
    14 -> 9 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    14 -> 33 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    14 -> 33 [ label = " 'cfg(any(target_os = \"dragonfly\", target_os = \"freebsd\", target_os = \"illumos\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    14 -> 15 [ label = " 'cfg(any(target_arch = \"x86\", target_arch = \"x86_64\", all(any(target_arch = \"aarch64\", target_arch = \"arm\"), any(target_os = \"android\", target_os = \"fuchsia\", target_os = \"linux\"))))'" ]
    14 -> 18 [ label = "" ]
    14 -> 34 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    14 -> 35 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    14 -> 36 [ label = " 'cfg(target_os = \"windows\")'" ]
    14 -> 37 [ label = " 'cfg(target_os = \"windows\")'" ]
    16 -> 12 [ label = "" ]
    16 -> 32 [ label = "" ]
    16 -> 13 [ label = "" ]
    16 -> 38 [ label = "" ]
    16 -> 17 [ label = "" ]
    19 -> 6 [ label = "" ]
    19 -> 21 [ label = "" ]
    19 -> 39 [ label = "" ]
    20 -> 40 [ label = "" ]
    20 -> 10 [ label = "" ]
    20 -> 41 [ label = "" ]
    20 -> 42 [ label = "" ]
    20 -> 43 [ label = "" ]
    20 -> 44 [ label = "" ]
    20 -> 45 [ label = "" ]
    20 -> 23 [ label = "" ]
    21 -> 43 [ label = "" ]
    21 -> 22 [ label = "" ]
    21 -> 46 [ label = "" ]
    22 -> 42 [ label = "" ]
    22 -> 43 [ label = "" ]
    22 -> 47 [ label = "" ]
    22 -> 44 [ label = "" ]
    22 -> 45 [ label = "" ]
    22 -> 20 [ label = "" ]
    22 -> 48 [ label = "" ]
    22 -> 23 [ label = "" ]
    24 -> 8 [ label = "" ]
    24 -> 31 [ label = "" ]
    25 -> 26 [ label = " 'i686-pc-windows-gnu'" ]
    25 -> 27 [ label = " 'x86_64-pc-windows-gnu'" ]
    40 -> 2 [ label = "" ]
    49 -> 3 [ label = "" ]
    50 -> 3 [ label = "" ]
    50 -> 49 [ label = "" ]
    51 -> 3 [ label = "" ]
    28 -> 3 [ label = "" ]
    28 -> 49 [ label = "" ]
    30 -> 7 [ label = "" ]
    33 -> 11 [ label = "" ]
    33 -> 52 [ label = "" ]
    53 -> 11 [ label = "" ]
    41 -> 11 [ label = "" ]
    41 -> 33 [ label = "" ]
    52 -> 11 [ label = "" ]
    52 -> 53 [ label = "" ]
    32 -> 12 [ label = "" ]
    42 -> 12 [ label = "" ]
    42 -> 32 [ label = "" ]
    38 -> 13 [ label = "" ]
    38 -> 32 [ label = "" ]
    43 -> 13 [ label = "" ]
    43 -> 38 [ label = "" ]
    54 -> 14 [ label = "" ]
    54 -> 11 [ label = "" ]
    55 -> 14 [ label = "" ]
    55 -> 54 [ label = "" ]
    29 -> 14 [ label = "" ]
    29 -> 56 [ label = "" ]
    29 -> 55 [ label = "" ]
    56 -> 14 [ label = "" ]
    47 -> 16 [ label = "" ]
    57 -> 16 [ label = "" ]
    57 -> 13 [ label = "" ]
    58 -> 16 [ label = "" ]
    58 -> 32 [ label = "" ]
    58 -> 38 [ label = "" ]
    59 -> 16 [ label = "" ]
    59 -> 57 [ label = "" ]
    60 -> 16 [ label = "" ]
    44 -> 16 [ label = "" ]
    61 -> 16 [ label = "" ]
    45 -> 16 [ label = "" ]
    45 -> 61 [ label = "" ]
    45 -> 60 [ label = "" ]
    45 -> 59 [ label = "" ]
    45 -> 62 [ label = "" ]
    45 -> 58 [ label = "" ]
    62 -> 16 [ label = "" ]
    63 -> 19 [ label = "" ]
    64 -> 19 [ label = "" ]
    64 -> 39 [ label = "" ]
    31 -> 19 [ label = "" ]
    31 -> 64 [ label = "" ]
    31 -> 63 [ label = "" ]
    48 -> 20 [ label = "" ]
    39 -> 21 [ label = "" ]
    39 -> 46 [ label = "" ]
    46 -> 22 [ label = "" ]
    46 -> 48 [ label = "" ]
    35 -> 24 [ label = "" ]
    35 -> 65 [ label = "" ]
    65 -> 24 [ label = "" ]
    34 -> 24 [ label = "" ]
    37 -> 25 [ label = "" ]
    36 -> 25 [ label = "" ]
}

